### PR TITLE
Added support for file share reliable download

### PR DESCRIPTION
**Pull request by gapra-msft**

### CODE REVIEW

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java [hunk 1]
```diff
@@ -826,25 +832,111 @@ private void channelCleanUp(AsynchronousFileChannel channel) {
...
+                            return Flux.empty();
+                        }
+
+                        try {
+                            return downloadRange(
+                                new ShareFileRange(offset, range.getEnd()), getRangeContentMd5,
+                                requestConditions, context).flatMapMany(r -> {
+                                    String receivedETag = ModelHelper.getETag(r.getHeaders());
+                                    if ((eTag == null && receivedETag == null)
+                                        || (eTag != null && eTag.equals(receivedETag))) {
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java:913]: should we assume etag isn't null? What would be scenario it is null ?

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ModelHelper.java [hunk 2]
```diff
@@ -191,6 +191,13 @@ public static ShareFileDownloadHeaders transformFileDownloadHeaders(HttpHeaders
         }
     }
 
+    public static String getETag(HttpHeaders headers) {
+        if (headers == null) {
+            return null;
+        }
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ModelHelper.java:197]: should we assume headers aren't null ?

**Review by kasobol-msft — COMMENTED** [review]
LGTM.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ModelHelper.java (hunk 2 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ModelHelper.java:197]: that seems like a fair assumption. Will remove.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java]: fixed

**Review by kasobol-msft — APPROVED** [review]

**Review by rickle-msft — APPROVED** [review]

**Pull request merged**